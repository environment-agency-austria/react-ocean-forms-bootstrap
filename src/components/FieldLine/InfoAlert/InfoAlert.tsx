/**
 * Copyright (c) 2018-present, Umweltbundesamt GmbH
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
import * as React from 'react';

import { FormText } from 'react-ocean-forms';
import { Alert } from 'reactstrap';

import { IInfoAlertProps } from './InfoAlert.types';

/**
 * Displays an bootstrap alert containing the parsed
 * info message
 */
export class InfoAlert extends React.Component<IInfoAlertProps> {
  public static displayName: string = 'InfoAlert';

  public render(): JSX.Element | null {
    const {
      visible,
      info,
      plaintext,
      onClose,
    } = this.props;

    if (plaintext || info === undefined) { return null; }

    return (
      <Alert
        color="success"
        className="mt-2"
        isOpen={visible}
        toggle={onClose}
      >
        <FormText text={info} />
      </Alert>
    );
  }
}
