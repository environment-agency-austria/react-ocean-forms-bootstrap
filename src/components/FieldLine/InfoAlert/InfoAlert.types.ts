/**
 * Props for the InfoAlert component
 */
export interface IInfoAlertProps {
  /**
   * True if the info alert should be visible
   */
  visible: boolean;
  /**
   * Info text
   */
  info?: string;
  /**
   * True if the form is in plaintext mode
   */
  plaintext: boolean;
  /**
   * onClose callback
   */
  onClose(): void;
}
