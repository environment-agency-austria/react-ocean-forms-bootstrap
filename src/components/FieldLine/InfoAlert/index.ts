export * from './InfoAlert';
export * from './InfoAlert.types';
