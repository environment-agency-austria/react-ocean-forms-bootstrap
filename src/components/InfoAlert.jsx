/**
 * Copyright (c) 2018-present, Umweltbundesamt GmbH
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
import React from 'react';
import PropTypes from 'prop-types';
import { Alert } from 'reactstrap';
import { fieldMetaShape } from 'react-ocean-forms';

/**
 * Displays an bootstrap alert containing the parsed
 * info message
 */
function InfoAlert({
  visible,
  info,
  meta,
  onClose,
}) {
  if (meta.plaintext) return null;
  if (!info) return null;

  const infoString = meta.stringFormatter(info);
  return (
    <Alert
      color="success"
      className="mt-2"
      isOpen={visible}
      toggle={onClose}
    >
      {infoString}
    </Alert>
  );
}

InfoAlert.defaultProps = {
  info: undefined,
};

InfoAlert.propTypes = {
  visible: PropTypes.bool.isRequired,
  info: PropTypes.string,
  meta: fieldMetaShape.isRequired,
  onClose: PropTypes.func.isRequired,
};

export default InfoAlert;
