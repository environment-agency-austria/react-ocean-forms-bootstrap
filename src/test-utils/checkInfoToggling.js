/**
 * Re-usable test suite to check the info visibility
 * toggling between InfoAlert and InfoAddonButton
 */
const checkInfoToggling = (wrapper) => {
  describe('info visible toggling', () => {
    const getVisibility = () => wrapper.find('InfoAlert').prop('visible');

    const toggleCases = [
      ['InfoAddonButton', 'onClick'],
      ['InfoAlert', 'onClose'],
    ];

    describe.each(toggleCases)('%s click', (element, prop) => {
      let visibility = getVisibility();

      it(`should toggle the info visibility on ${element} click`, () => {
        for (let i = 0; i < 2; i += 1) {
          wrapper.find(element).prop(prop)();
          const newVisibility = getVisibility();
          expect(newVisibility).toBe(!visibility);
          visibility = newVisibility;
        }
      });
    });
  });
};

export default checkInfoToggling;
