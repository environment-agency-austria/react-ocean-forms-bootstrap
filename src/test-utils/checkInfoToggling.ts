import { ShallowWrapper } from 'enzyme';

/**
 * Re-usable test suite to check the info visibility
 * toggling between InfoAlert and InfoAddonButton
 */
export const checkInfoToggling = (wrapper: ShallowWrapper): void => {
  describe('info visible toggling', () => {
    const getVisibility = (): boolean => wrapper.find('InfoAlert').prop('visible');

    const toggleCases = [
      ['InfoAddonButton', 'onClick'],
      ['InfoAlert', 'onClose'],
    ];

    describe.each(toggleCases)('%s click', (element: string, prop: string) => {
      let visibility = getVisibility();

      it(`should toggle the info visibility on ${element} click`, () => {
        for (let i = 0; i < 2; i += 1) {
          const callback: Function = wrapper.find(element).prop(prop);
          callback();

          const newVisibility = getVisibility();
          expect(newVisibility).toBe(!visibility);
          visibility = newVisibility;
        }
      });
    });
  });
};
