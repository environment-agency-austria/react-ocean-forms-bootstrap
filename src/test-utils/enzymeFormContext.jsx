/**
 * Copyright (c) 2018-present, Umweltbundesamt GmbH
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import createMockFormatter from './createMockFormatter';

/**
 * Components inside the form module require access to the form context.
 * This is not available when mounting single components in Enzyme.
 * These helper functions aim to address that and wrap a valid,
 * user definable form context around them.
 */

/**
 * Creates a form context
 */
export const createMockFormContext = registerCallback => ({
  fieldPrefix: null,

  registerField: jest.fn().mockImplementation((name, state) => registerCallback(name, state)),
  unregisterField: jest.fn(),
  notifyFieldEvent: jest.fn(),

  registerListener: jest.fn().mockImplementation((name, state) => registerCallback(name, state)),
  unregisterListener: jest.fn(),

  getFieldState: jest.fn(),
  getValues: jest.fn(),

  stringFormatter: createMockFormatter(),

  busy: false,
  disabled: false,
  plaintext: false,

  asyncValidateOnChange: false,
  asyncValidationWait: 400,
  defaultValues: {},
});

/**
 * Creates a mock field meta
 */
export const createMockFieldMeta = () => ({
  valid: true,
  error: null,
  isValidating: false,
  touched: false,
  stringFormatter: createMockFormatter(),
  plaintext: false,
});

/**
 * Creates a mock field shape
 */
export const createMockField = () => ({
  id: 'mock-field',
  name: 'mock-field',
  value: 'mock-value',
  disabled: false,
  onChange: jest.fn(),
  onBlur: jest.fn(),
});
