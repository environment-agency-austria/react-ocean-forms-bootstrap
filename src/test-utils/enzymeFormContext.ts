/**
 * Copyright (c) 2018-present, Umweltbundesamt GmbH
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
import { IFieldComponentFieldProps, IFieldComponentMeta, IFieldState, IFormContext, TFormEventListener } from 'react-ocean-forms';

import { createMockFormatter } from './createMockFormatter';

/**
 * Components inside the form module require access to the form context.
 * This is not available when mounting single components in Enzyme.
 * These helper functions aim to address that and wrap a valid,
 * user definable form context around them.
 */

/**
 * Creates a form context
 */
export const createMockFormContext = (registerCallback?: Function): IFormContext => ({
  fieldPrefix: null,

  registerField: jest.fn().mockImplementation((name: string, state: IFieldState): void => {
    if (registerCallback) {
      registerCallback(name, state);
    }
  }),
  unregisterField: jest.fn(),
  notifyFieldEvent: jest.fn(),

  registerListener: jest.fn().mockImplementation((name: string, callback: TFormEventListener): void => {
    if (registerCallback) {
      registerCallback(name, callback);
    }
  }),
  unregisterListener: jest.fn(),

  getFieldState: jest.fn(),
  getValues: jest.fn(),

  stringFormatter: createMockFormatter(),
  submit: jest.fn(),

  busy: false,
  disabled: false,
  plaintext: false,

  asyncValidateOnChange: false,
  asyncValidationWait: 400,
  defaultValues: {},
});

/**
 * Creates a mock field meta
 */
export const createMockFieldMeta = (): IFieldComponentMeta => ({
  valid: true,
  error: null,
  isValidating: false,
  isRequired: false,
  touched: false,
  stringFormatter: createMockFormatter(),
  plaintext: false,
});

/**
 * Creates a mock field shape
 */
export const createMockField = (): IFieldComponentFieldProps => ({
  id: 'mock-field',
  name: 'mock-field',
  value: 'mock-value',
  disabled: false,
  onChange: jest.fn(),
  onBlur: jest.fn(),
});
