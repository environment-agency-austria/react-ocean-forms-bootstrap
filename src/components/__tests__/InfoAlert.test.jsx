import React from 'react';
import { shallow } from 'enzyme';

import InfoAlert from '../InfoAlert';
import { createMockFieldMeta } from '../../test-utils/enzymeFormContext';

describe('<InfoAlert />', () => {
  const setup = (plaintext, info, visible = false, onClose = () => {}) => {
    const meta = createMockFieldMeta();
    meta.plaintext = plaintext;

    const wrapper = shallow((
      <InfoAlert
        meta={meta}
        visible={visible}
        info={info}
        onClose={onClose}
      />
    ));

    return { wrapper, meta };
  };

  it('should render nothing if meta.plaintext is active', () => {
    const { wrapper } = setup(true);
    expect(wrapper.exists('Alert')).toBeFalsy();
  });

  it('should render nothing if there is no info', () => {
    const { wrapper } = setup(false, null);
    expect(wrapper.exists('Alert')).toBeFalsy();
  });

  describe('info present', () => {
    const MOCK_INFO = 'mock-info';
    const MOCK_VISIBLE = true;
    const MOCK_ON_CLOSE = jest.fn();

    const { wrapper, meta } = setup(false, MOCK_INFO, MOCK_VISIBLE, MOCK_ON_CLOSE);

    it('should run the info through meta.stringFormatter', () => {
      expect(meta.stringFormatter).toHaveBeenCalledWith(MOCK_INFO);
    });

    it('should render an Alert with the correct props', () => {
      expect(wrapper.exists('Alert')).toBeTruthy();

      const alert = wrapper.find('Alert');
      expect(alert.prop('isOpen')).toBe(MOCK_VISIBLE);
      expect(alert.prop('toggle')).toBe(MOCK_ON_CLOSE);
    });

    it('should render correctly', () => {
      expect(wrapper).toMatchSnapshot();
    });
  });
});
