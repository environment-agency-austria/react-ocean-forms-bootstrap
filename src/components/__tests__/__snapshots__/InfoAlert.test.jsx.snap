// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<InfoAlert /> info present should render correctly 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <InfoAlert
    info="mock-info"
    meta={
      Object {
        "error": null,
        "isValidating": false,
        "plaintext": false,
        "stringFormatter": [MockFunction] {
          "calls": Array [
            Array [
              "mock-info",
            ],
          ],
          "results": Array [
            Object {
              "isThrow": false,
              "value": "mock-info",
            },
          ],
        },
        "touched": false,
        "valid": true,
      }
    }
    onClose={[MockFunction]}
    visible={true}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "children": "mock-info",
      "className": "mt-2",
      "closeAriaLabel": "Close",
      "color": "success",
      "fade": true,
      "isOpen": true,
      "tag": "div",
      "toggle": [MockFunction],
      "transition": Object {
        "appear": true,
        "baseClass": "fade",
        "baseClassActive": "show",
        "enter": true,
        "exit": true,
        "in": true,
        "mountOnEnter": false,
        "onEnter": [Function],
        "onEntered": [Function],
        "onEntering": [Function],
        "onExit": [Function],
        "onExited": [Function],
        "onExiting": [Function],
        "tag": "div",
        "timeout": 150,
        "unmountOnExit": true,
      },
    },
    "ref": null,
    "rendered": "mock-info",
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": "mock-info",
        "className": "mt-2",
        "closeAriaLabel": "Close",
        "color": "success",
        "fade": true,
        "isOpen": true,
        "tag": "div",
        "toggle": [MockFunction],
        "transition": Object {
          "appear": true,
          "baseClass": "fade",
          "baseClassActive": "show",
          "enter": true,
          "exit": true,
          "in": true,
          "mountOnEnter": false,
          "onEnter": [Function],
          "onEntered": [Function],
          "onEntering": [Function],
          "onExit": [Function],
          "onExited": [Function],
          "onExiting": [Function],
          "tag": "div",
          "timeout": 150,
          "unmountOnExit": true,
        },
      },
      "ref": null,
      "rendered": "mock-info",
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getSnapshotBeforeUpdate": true,
        },
      },
    },
    "attachTo": undefined,
    "hydrateIn": undefined,
  },
}
`;
